### PR TITLE
Fix signals to update ES index

### PR DESCRIPTION
**Pull request by saruniitr**
## Change description

Previous changes to update the ES index on data updates are not working.
The registry expects model corresponding to the index which is BaseApplication
in our case so the updates for related models were not happening. Update
signals to get the relevant BaseApplication instance.

### CODE REVIEW

File: api/search/signals.py [hunk 1]
```diff
@@ -1,41 +1,59 @@
...
-            registry.update(instance)
+    """
+    To keep the index fresh we need to update them whenever the underlying data is updated.
+    The immediate attributes of the index are handled by django_elasticsearch_dsl but
+    additional changes are required to update the related models.
 
+    The registry update code requires model corresponding to the index which is BaseApplication
+    in our case. We intercept the signals and update the index.
+    """
+    if app_label == "applications":
```
Comment (richtier) [inline api/search/signals.py:23]: would this do the same? `if issubclass(instance, BaseApplication) or issubclass(instance, BaseApplication):`

File: api/search/signals.py [hunk 2]
```diff
@@ -1,41 +1,59 @@
...
+        if model_name in [
+            "standardapplication",
+            "baseapplication",
+            "openapplication",
+            "exhibitionclearanceapplication",
+        ]:
+            application = BaseApplication.objects.get(id=instance.id)
```
Comment (richtier) [inline api/search/signals.py:30]: `application = instance.baseapplication`

File: api/search/signals.py [hunk 3]
```diff
@@ -1,41 +1,59 @@
...
+                registry.update(poa.application)
 
     if app_label == "organisations":
         if model_name == "organisation":
-            instances = instance.cases.all()
-            for _instance in instances:
-                registry.update(_instance)
+            for case in instance.cases.all():
+                try:
+                    base_application_id = BaseApplication.objects.get(id=case.id)
```
Comment (richtier) [inline api/search/signals.py:56]: `base_application_id` implies the variable is a primary key, but it's really and instance of BaseApplication

Comment (saruniitr) [reply api/search/signals.py:56]: indeed, will update this.

File: api/search/signals.py (hunk 2 above)
Comment (saruniitr) [reply api/search/signals.py:30]: ok

File: api/search/signals.py (hunk 1 above)
Comment (saruniitr) [reply api/search/signals.py:23]: yes that is better

File: api/search/signals.py [hunk 4]
```diff
@@ -1,41 +1,52 @@
...
+    To keep the index fresh we need to update them whenever the underlying data is updated.
+    The immediate attributes of the index are handled by django_elasticsearch_dsl but
+    additional changes are required to update the related models.
 
-        if model_name == "goodonapplication":
-            registry.update(instance)
+    The registry update code requires model corresponding to the index which is BaseApplication
+    in our case. We intercept the signals and update the index.
+    """
+    if issubclass(instance.__class__, BaseApplication):
```
Comment (richtier) [inline api/search/signals.py:23]: I think `sender` is the instance's class

File: api/search/signals.py [hunk 5]
```diff
@@ -1,41 +1,52 @@
...
+                registry.update(poa.application)
 
     if app_label == "organisations":
         if model_name == "organisation":
-            instances = instance.cases.all()
-            for _instance in instances:
-                registry.update(_instance)
+            for case in instance.cases.all():
+                try:
+                    base_application = BaseApplication.objects.get(id=case.id)
```
Comment (richtier) [inline api/search/signals.py:49]: `case.baseapplication` works here too. When a model is subclassed, django adds a pointer on the subclass to the superclass (with the superclass's app label as the attribute name) 

so it works on the `caseassignment` case too

Comment (saruniitr) [reply api/search/signals.py]: fixed

File: api/search/signals.py (hunk 4 above)
Comment (saruniitr) [reply api/search/signals.py]: updated

**Review by saruniitr — COMMENTED** [review]
+1 for #623 changes

**Pull request merged**